components/raftstore-v2/src/operation/command/admin/merge/prepare.rs

```diff
@@ -109,10 +125,18 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             store_ctx,
             req.get_admin_request().get_prepare_merge(),
         )?;
+        // We need to check three things in order:
+        // (1) `check_logs_before_prepare_merge`
+        // (2) `check_trim_status`
+        // (3) `check_pessimistic_locks`
+        // Check 2 and 3 are async, they yield by returning error.
         let pre_propose = if let Some(r) = self.already_checked_pessimistic_locks()? {
             r
+        } else if let Some(r) = self.already_checked_trim_status()? {
+            self.check_pessimistic_locks(r, &mut req)?
```

Comment:
What if a conf change is proposed right after `check_pessimistic_locks`?

Reply:
Then the merge will be rejected? Because epoch is changed.

Reply:
What if a follower is isolated right after `check_trim_status` and leader proposes a lot of logs before `already_checked_trim_status`?

Reply:
Leader will timeout and drop the prepare merge request.

Reply:
What if the trim response is sent, delayed and finarlly delivered after some time?

Reply:
It will be ignored if either there's no pending prepare merge, or the epoch doesn't match the pending request.

Reply:
What if there are still on going pending prepare merge and epoch is the same? I can see the timeout is 1s. Nothing can prevent leader propose an admin command (and not apply yet) or even propose hundreds of proposals during that time.

Reply:
The result of `check_logs_before_prepare_merge` becomes unreliable, anything can happen. For example, a compact log (as well as admin commands that changes epoch) proposed during this time can break the merge algorithm.